src/analysis/locally_convex/basic.lean

```diff
@@ -137,61 +137,69 @@ forall₂_congr $ λ a ha, smul_set_subset_iff
 
 alias balanced_iff_smul_mem ↔ balanced.smul_mem _
 
-lemma balanced_univ : balanced 𝕜 (univ : set E) := λ a ha, subset_univ _
+@[simp] lemma balanced_empty : balanced 𝕜 (∅ : set E) :=
+λ _ _, by { rw smul_set_empty }
+
+@[simp] lemma balanced_univ : balanced 𝕜 (univ : set E) := λ a ha, subset_univ _
 
 lemma balanced.union (hA : balanced 𝕜 A) (hB : balanced 𝕜 B) : balanced 𝕜 (A ∪ B) :=
-begin
-  intros a ha t ht,
-  rw smul_set_union at ht,
-  exact ht.imp (λ x, hA _ ha x) (λ x, hB _ ha x),
-end
+λ a ha, smul_set_union.subset.trans $ union_subset_union (hA _ ha) $ hB _ ha
 
 lemma balanced.inter (hA : balanced 𝕜 A) (hB : balanced 𝕜 B) : balanced 𝕜 (A ∩ B) :=
-begin
-  rintro a ha _ ⟨x, ⟨hx₁, hx₂⟩, rfl⟩,
-  exact ⟨hA _ ha ⟨_, hx₁, rfl⟩, hB _ ha ⟨_, hx₂, rfl⟩⟩,
-end
+λ a ha, smul_set_inter_subset.trans $ inter_subset_inter (hA _ ha) $ hB _ ha
+
+lemma balanced_Union {f : ι → set E} (h : ∀ i, balanced 𝕜 (f i)) : balanced 𝕜 (⋃ i, f i) :=
+λ a ha, (smul_set_Union _ _).subset.trans $ Union_mono $ λ _, h _ _ ha
+
+lemma balanced_Union₂ {f : Π i, κ i → set E} (h : ∀ i j, balanced 𝕜 (f i j)) :
+  balanced 𝕜 (⋃ i j, f i j) :=
+balanced_Union $ λ _, balanced_Union $ h _
+
+lemma balanced_Inter {f : ι → set E} (h : ∀ i, balanced 𝕜 (f i)) : balanced 𝕜 (⋂ i, f i) :=
+λ a ha, (smul_set_Inter_subset _ _).trans $ Inter_mono $ λ _, h _ _ ha
+
+lemma balanced_Inter₂ {f : Π i, κ i → set E} (h : ∀ i j, balanced 𝕜 (f i j)) :
+  balanced 𝕜 (⋂ i j, f i j) :=
+balanced_Inter $ λ _, balanced_Inter $ h _
+
+variables [has_scalar 𝕝 E] [smul_comm_class 𝕜 𝕝 E]
+
+lemma balanced.smul (a : 𝕝) (hs : balanced 𝕜 s) : balanced 𝕜 (a • s) :=
+λ b hb, (smul_comm _ _ _).subset.trans $ smul_set_mono $ hs _ hb
 
 end has_scalar
 
-section add_comm_monoid
-variables [add_comm_monoid E] [module 𝕜 E] {s s' t t' u v A B : set E}
+section module
+variables [add_comm_group E] [module 𝕜 E] {s s₁ s₂ t t₁ t₂ : set E}
 
-lemma absorbs.add (h : absorbs 𝕜 s t) (h' : absorbs 𝕜 s' t') : absorbs 𝕜 (s + s') (t + t') :=
-begin
-  rcases h with ⟨r, hr, h⟩,
-  rcases h' with ⟨r', hr', h'⟩,
-  refine ⟨max r r', lt_max_of_lt_left hr, λ a ha, _⟩,
-  rw smul_add,
-  exact set.add_subset_add (h a (le_of_max_le_left ha)) (h' a (le_of_max_le_right ha)),
-end
+lemma absorbs.neg : absorbs 𝕜 s t → absorbs 𝕜 (-s) (-t) :=
+Exists.imp $ λ r, and.imp_right $ forall₂_imp $ λ _ _ h,
+  (neg_subset_neg.2 h).trans set.smul_set_neg.superset
 
-lemma balanced.add (hA₁ : balanced 𝕜 A) (hA₂ : balanced 𝕜 B) : balanced 𝕜 (A + B) :=
-begin
-  rintro a ha _ ⟨_, ⟨x, y, hx, hy, rfl⟩, rfl⟩,
-  rw smul_add,
-  exact add_mem_add (hA₁ _ ha ⟨_, hx, rfl⟩) (hA₂ _ ha ⟨_, hy, rfl⟩),
-end
+lemma balanced.neg : balanced 𝕜 s → balanced 𝕜 (-s) :=
+forall₂_imp $ λ _ _ h, smul_set_neg.subset.trans $ neg_subset_neg.2 h
 
-lemma balanced_zero : balanced 𝕜 (0 : set E) := λ a ha, (smul_zero _).subset
+lemma absorbs.add : absorbs 𝕜 s₁ t₁ → absorbs 𝕜 s₂ t₂ → absorbs 𝕜 (s₁ + s₂) (t₁ + t₂) :=
```

Comment:
Isn't this now less general than before? If I'm reading the diff right, it previously required `[add_comm_monoid E]` and now we need `[add_comm_group E]`. (Same for `balanced.add`.)

Reply:
Yes exactly. We have negation on `E` anyway, so it's better to assume it in the typeclasses to reduce the number of combinations of instance assumptions and avoid people trying to "generalize" stuff meaninglessly.

Reply:
Fair enough!